* challenge sample definitions didn't work out well for long_short:
see tests/complete/input/iso_detect_de_novo_no_prep
since the lists of prep/platforms are independent, it want to
require manatee block, because of ES cDNA/Illumina, but it didn't
have dRNA.

entry_md long_short [<LibraryPrep.cDNA: 4>, <LibraryPrep.dRNA: 2>] [<Platform.Illumina: 1>, <Platform.ONT: 3>]
long_runs {(<Sample.ES: 3>, <LibraryPrep.cDNA: 4>, <Platform.ONT: 3>),
           (<Sample.ES: 3>, <LibraryPrep.dRNA: 2>, <Platform.ONT: 3>),
           (<Sample.blood: 4>, <LibraryPrep.cDNA: 4>, <Platform.ONT: 3>)}
short_runs {(<Sample.blood: 4>, <LibraryPrep.cDNA: 4>, <Platform.Illumina: 1>),
            (<Sample.ES: 3>, <LibraryPrep.cDNA: 4>, <Platform.Illumina: 1>)}
long_samples {<Sample.ES: 3>, <Sample.blood: 4>}
short_samples {<Sample.ES: 3>, <Sample.blood: 4>}

* ENCODE search
https://www.encodeproject.org/search/?type=Experiment&internal_tags=LRGASP
* DOC:
** FAQ
** step-by-step
** better data overview
** QC tables
** self-evaluation
* validate
** check that expression matrix contains all replicates and validate ids
for paired-end, take either id
** 1&2n all the above categories, the genome and transcriptome references specified by LRGASP should be used. For the long and short and freestyle category, additional transcriptome references can be used.
** 3 In all the above categories, except for freestyle a transcriptome reference CANNOT be used.
** create command line script to download all files
** entry is against all samples
** only one freestyle
** short-only - no read-map
** long-and-short additional accessioned data in public genomics data repositories can also be used. [in slides]

* freestyle must be public at the start of the challenge available
* SUBMISSION IMAGE
** code broken
** doesn't show model for quant
** should show denovo
* issues:
** read_model_map for short-read only experiment?
* metadata need
manatee
simulation
* improve doc library prep method challenges.md
* mention they must do their own alignments
* fix example files software to match experiment types
* why does run ENCSR507JOF have multiple bams in same replicate
* what are the tar files in ENCSR392BGY?
* need to check for duplicate files from same replicate
* encode download cmd is
xargs -L 1 curl -O -J -L < files.txt
* need to make reference genome that includes the SIRVs
https://www.encodeproject.org/references/ENCSR089MWE/
* remove reference-genomes.md
* read-model map says: Additionally, participants may want to submit a BED12 format to map the initial sequences to the reference genome. To provide or not this information is up to the submitters because this might be a bit complicated with some pipelines.

* have lrgasp-validate-entry print each entry being validated
* fix software descriptions to be better
* make extra_libraries optional
* https://docs.google.com/presentation/d/1RxmZPd8UgWgm3F-0BquP7NmhxE3zYB5lJ_u6XKgbFME/edit#slide=id.p
* require protocol description??
* add assembly
* doc that file paths are relative
* add additional file types
* should expression matrix have transcript_id instead of ID?
* require included models_gtf in quant.
* check experiment_type is deduced from challenge id
challenge_id can be obtained from entry_id (add to experiment)
** validate same type of experiments
* update refgenomes doc
