* issues:
** TPM only
** read_model_map for short-read only experiment?
* remove lost of info with select catchs:
except (LrgaspException, ParsingError, ValueError) as ex:
        raise GtfException("Parse of GTF failed: {}".format(model_gtf)) from ex

* metadata need
manatee
simulation
* bad SIRV genes
** SIRVs 1-6 have + and - strand “transcripts”
** SIRV107has 
SIRV107	unknownUtrSplice	CT..AG	SIRV1	11242	11403
SIRV107	unknownUtrSplice	CT..AC	SIRV1	10791	10882
* improve doc library prep method challenges.md
* mention they must do their own alignments
* fix example files software to match experiment types
* why does run ENCSR507JOF have multiple bams in same replicate
* what are the tar files in ENCSR392BGY?
* need to check for duplicate files from same replicate
* encode download cmd is
xargs -L 1 curl -O -J -L < files.txt
* need to make reference genome that includes the SIRVs
https://www.encodeproject.org/references/ENCSR089MWE/
* remove reference-genomes.md
* read-model map says: Additionally, participants may want to submit a BED12 format to map the initial sequences to the reference genome. To provide or not this information is up to the submitters because this might be a bit complicated with some pipelines.

* have lrgasp-validate-entry print each entry being validated
* fix software descriptions to be better
* make extra_libraries optional
* is ES really the right library name for mm??
* validate libraries, extra_libraries, is_kitchen_sink, etc, are valid and consistent
* https://docs.google.com/presentation/d/1RxmZPd8UgWgm3F-0BquP7NmhxE3zYB5lJ_u6XKgbFME/edit#slide=id.p
* require protocol description??
* add assembly
* doc that file paths are relative
* add additional file types
* should expression matrix have transcript_id instead of ID?
* require included model_gtf in quant.
* check experiment_type is deduced from challange id
challange_id can be obtained from entry_id (add to experiment)
** validate same type of experiments
* update refgenomes doc
