* update ENCODE metadata
* DOC:
** FAQ
** better data overview
** QC tables
** self-evaluation
* validate
** 1&2n all the above categories, the genome and transcriptome references specified by LRGASP should be used. For the long and short and kitchen sink category, additional transcriptome references can be used.
** 3 In all the above categories, except for kitchen sink a transcriptome reference CANNOT be used.
** transcriptome 
** create command line script to download all files
** entry is against all samples
** insure sizes select and non-size select?
** only one kitchen sink
** short-only - no read-map
** long-and-short additional accessioned data in public genomics data repositories can also be used. [in slides]

** does ch #3 require all samples?
** only one entry per data category + library type
* kitchen sink must be public at the start of the challange available
* add manatee genome
* SUBMISSION IMAGE
** code broken
** doesn't show model for quant
** should show denovo
* issues:
** read_model_map for short-read only experiment?
* metadata need
manatee
simulation
* improve doc library prep method challenges.md
* mention they must do their own alignments
* fix example files software to match experiment types
* why does run ENCSR507JOF have multiple bams in same replicate
* what are the tar files in ENCSR392BGY?
* need to check for duplicate files from same replicate
* encode download cmd is
xargs -L 1 curl -O -J -L < files.txt
* need to make reference genome that includes the SIRVs
https://www.encodeproject.org/references/ENCSR089MWE/
* remove reference-genomes.md
* read-model map says: Additionally, participants may want to submit a BED12 format to map the initial sequences to the reference genome. To provide or not this information is up to the submitters because this might be a bit complicated with some pipelines.

* have lrgasp-validate-entry print each entry being validated
* fix software descriptions to be better
* make extra_libraries optional
* is ES really the right library name for mm??
* validate libraries, extra_libraries, is_kitchen_sink, etc, are valid and consistent
* https://docs.google.com/presentation/d/1RxmZPd8UgWgm3F-0BquP7NmhxE3zYB5lJ_u6XKgbFME/edit#slide=id.p
* require protocol description??
* add assembly
* doc that file paths are relative
* add additional file types
* should expression matrix have transcript_id instead of ID?
* require included model_gtf in quant.
* check experiment_type is deduced from challange id
challange_id can be obtained from entry_id (add to experiment)
** validate same type of experiments
* update refgenomes doc
