---
title: "Evaluation metrics LRGASP"
output: html_document
params:
  date: !r Sys.Date()
  output.directory: ""   #folder where the evaluation output will be saved
  Name: "Example"
  Platform: "PacBio"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
getwd()
```

# Evaluation metrics LRGASP

```{r, echo=FALSE}

## These are a few libraries that must be loaded for the plots and also a couple of variables to make them "nice"

suppressWarnings(library(ggplot2))
suppressWarnings(library(scales))
suppressWarnings(library(kableExtra))

myPalette = c("#6BAED6","#FC8D59","#78C679","#EE6A50","#969696","#66C2A4", "goldenrod1", "darksalmon", "#41B6C4","tomato3", "#FE9929")

mytheme <- theme_classic(base_family = "Helvetica") +
  theme(axis.line.x = element_line(color="black", size = 0.4),
        axis.line.y = element_line(color="black", size = 0.4)) +
  theme(axis.title.x = element_text(size=14),
        axis.text.x  = element_text(size=13),
        axis.title.y = element_text(size=14),
        axis.text.y  = element_text(vjust=0.5, size=13) ) +
  theme(legend.text = element_text(size = 10), legend.title = element_text(size=11), legend.key.size = unit(0.5, "cm")) +
  theme(plot.title = element_text(lineheight=.4, size=13)) +
  theme(plot.margin = unit(c(2.5,1,1,1), "cm")) 

```

```{r, echo=FALSE}
setwd(params$output.directory)

results_file <- dir(pattern = "*_results.RData")
load(results_file)
names(all.results) <- c("FSM", "ISM", "NIC", "NNC", "SIRV")

FSM_class_file=dir(pattern = "*_FSM.RData")
load(FSM_class_file)
ISM_class_file=dir(pattern = "*_ISM.RData")
load(ISM_class_file)
NIC_class_file=dir(pattern = "*_NIC.RData")
load(NIC_class_file)
NNC_class_file=dir(pattern = "*_NNC.RData")
load(NNC_class_file)
SIRV_class_file=dir(pattern = "*_SIRVs_class.RData")
load(SIRV_class_file)

``` 


## Metrics description

As there is no ground truth for the cell-line transcriptome, we will call an isoform True Positive (TP) if there is external support for its 5' and/or 3' ends. For example, we will consider TP a FSM isoform if **both** of its ends are 50bp or closer to the annotated ends of the associated transcript. 

In the case of *Associated gene* TP ratios, we will take into account not only the 5' and 3' ends of the associated transcript, but also all the TSS and TTS annotated for that gene even if they are related to a different isoform. *All TP rates* use all the evidence available (annotated CAGE sites, polyA motif found and gene TTS/TSS) and we will call TP when there is any type of support to both of the ends. All these definitions are in the metrics Proposal LRGASP pdf that Ana Conesa prepared. These metrics are complementary to the pdf report generated by SQANTI3


### Evaluation of FSM

```{r , echo=FALSE}

kable(all.results["FSM"], align = "r")

ggplot(sqanti_data_FSM, aes(x=log(mean_all_coverage+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette[1])+
  labs(title="Mean SJ coverage",
       subtitle = "FSM isoforms") +
  xlab("log(Mean coverage of SJ +1)")

```

### Evaluation of ISM

```{r , echo=FALSE}

kable(all.results["ISM"], align = "r")
ggplot(sqanti_data_ISM, aes(x=log(mean_all_coverage+1), fill=structural_category ) )+
  geom_density(na.rm = T) +
  mytheme +
  scale_fill_manual(values=myPalette[2])+
  labs(title="Mean SJ coverage",
       subtitle = "ISM isoforms") +
  xlab("log(Mean coverage of SJ +1)")

ggplot(sqanti_data_ISM, aes( x=subcategory, y=missing_exons_perc, fill=subcategory)) +
  geom_boxplot() + 
  mytheme + 
  scale_fill_discrete()+
  labs(title = "Percentage of missing exons regarding to the associated reference transcript",
       subtitle = "The higher the % is, the more exons are missing") +
  ylab("Percentage of missing exons") +
  xlab("Subcategories") +
  coord_flip()

```

## Evaluation NIC

```{r , echo=FALSE}

kable(all.results["NIC"], align = "r")

ggplot(sqanti_data_NIC) +
  geom_density(color="black", size=0.5, aes( x = log(mean_novel_coverage +1) , fill="Just novel SJ"), alpha=0.5, na.rm = T) +
  geom_density(color="black",  size=0.5, aes( x = log(mean_all_coverage +1) , fill="All SJ"), alpha=0.5, na.rm = T) +
  mytheme  +
  labs(title="Mean SJ coverage per isoform comparison: all SJ vs novel SJ",
       subtitle = "Using all NIC isoforms") +
  ylab("Density") + xlab("log(Mean Coverage of SJ +1)") 

nic_only=subset(sqanti_data_NIC, FSM_class=="A")
  
ggplot(nic_only) +
  geom_density(color="black", size=0.5, aes( x = log(mean_novel_coverage +1) , fill="Just novel SJ"), alpha=0.5, na.rm = T) +
  geom_density(color="black",  size=0.5, aes( x = log(mean_all_coverage +1) , fill="All SJ"), alpha=0.5, na.rm = T) +
  mytheme  +
  labs(title="Mean SJ coverage per isoform comparison: all SJ vs novel SJ",
       subtitle = "Using just NIC isoforms which are the only transcript expressed for that gene") +
  ylab("Density") + xlab("log(Mean Coverage of SJ +1)")

```


## Evaluation NNC

```{r, echo=FALSE}

kable(all.results["NNC"], align = "r")

ggplot(sqanti_data_NNC) +
  geom_density(color="black", size=0.5, aes( x = log(mean_novel_coverage +1) , fill="Just novel SJ"), alpha=0.5, na.rm = T) +
  geom_density(color="black",  size=0.5, aes( x = log(mean_all_coverage +1) , fill="All SJ"), alpha=0.5, na.rm = T) +
  mytheme  +
  labs(title="Mean SJ coverage per isoform comparison: all SJ vs novel SJ",
       subtitle = "Using all NNC isoforms") +
  ylab("Density") + xlab("log(Mean Coverage of SJ +1)") 

nnc_only=subset(sqanti_data_NNC, (FSM_class=="A"))
  
ggplot(nnc_only) +
  geom_density(color="black", size=0.5, aes( x = log(mean_novel_coverage +1) , fill="Just novel SJ"), alpha=0.5, na.rm = T) +
  geom_density(color="black",  size=0.5, aes( x = log(mean_all_coverage +1) , fill="All SJ"), alpha=0.5, na.rm = T) +
  mytheme  +
  labs(title="Mean SJ coverage per isoform comparison: all SJ vs novel SJ",
       subtitle = "Using just NNC isoforms which are the only transcript expressed for that gene") +
  ylab("Density") + xlab("log(Mean Coverage of SJ +1)")



```


## Evaluation of Spike-Ins (SIRVs) 

Here we will evaluate if the transcriptome contains the SIRVs that should be detected, how many are recovered perfectly and how many False Positives it is reporting. We consider TP a FSM of an expected SIRV with its TSS and TTS closer than 50bp of the annotated ones. False Psitives are anything that is not FSM. Finally, FN will be the number of missing SIRVs. SIRVs Set3 contains 69 spliced isoforms.

We will also plot the expression values of the TP SIRVs. If the quantification is perfect, they should have the exact same expression values because they were spiked-in at equimolar concentrations.

**ATTENTION**
If in this chunk of the evaluation all the results are 0, please, check if the reference genome and/or transcriptome used for building your transcript-model are contain information about spike-ins.


```{r , echo=FALSE}
kable(all.results["SIRV"], align = "r")

TP_sirvs=sirv_data[which(sirv_data$structural_category=="full-splice_match" & 
                                           abs(sirv_data$diff_to_TSS)<=50 & abs(sirv_data$diff_to_TTS)<=50),]

TP_sirvs$log_exp=log(TP_sirvs$iso_exp+1)
ggplot(TP_sirvs, aes(x=associated_transcript, y=log_exp))  +
  geom_point() +
  mytheme + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1, size=8)) +
  scale_fill_discrete()+
  labs(title = "Expression values for the TP SIRVs ",
       subtitle = "If several dots appear for the same SIRV that's because of redundancy detecting SIRVs") +
  ylab("log(Expression value +1)") +
  xlab("SIRVs")

```


